Simplify defining access to resources for principals with several capabilities

skuenzli opened the pull request.

brentryan commented on lib/k9policy.ts:
```diff
@@ -18,7 +18,8 @@ export enum AccessCapability {
 export interface AccessSpec {
-    accessCapability: AccessCapability
+    accessCapability?: AccessCapability
+    accessCapabilities?: Set<AccessCapability>
```
I wonder if you could simplify this interface by doing:
```
accessCapabilities?: Set<AccessCapability> | AccessCapability
```

I could go either way, but thinking about other styles/patterns I've seen.

skuenzli replied:
Thanks for your feedback!

I started with a single `accessCapability` or `-ilities` (preferred) attribute.

But TypeScript erases all type information at runtime and so you can't do e.g. `accessSpec.accessCapabilities instanceof AccessCapability` to dispatch to the right code:
https://stackoverflow.com/questions/33800497/check-if-an-object-implements-an-interface-at-runtime-with-typescript

So a single attribute that accepts multiple types results in 'consumers' having to detect the object type by looking for properties like a `Set`'s `entries`.  I went down that path and started to become concerned that we'd need `isAccessCapability` and `isAccessCapabilities` helper functions to do this correctly.  Which is probably fine for k9-cdk.  But what about client code?  They're going to have to handle this complecting, too.  So I went down what looked like the simplest path.

That said, having both an `accessCapability` and `accessCapabilities` attribute feels wrong.

Thinking of deprecating the single `accessCapability` attribute and removing it after people have a chance to transition to `accessCapabilities`.

wdyt?

brentryan replied:
I've seen this done where as an input you allow an array input or a singular input but when you return the data, it's always an array. The singular input is just there for convenience as a client consumes the api. 

I'll see if I can find other examples as a reference.

brentryan replied:
Basically like the accepted answer here https://stackoverflow.com/questions/35749833/typescript-function-taking-one-or-array-of-objects/35750406

This operates on an array instead of a Set. Not sure that really matters though.

skuenzli commented:
Ok - I think we've got it.  I've updated `AccessSpec` to the model you suggested:
```
export interface AccessSpec {
    accessCapabilities: Set<AccessCapability> | AccessCapability
    allowPrincipalArns: Set<string>
    test?: ArnConditionTest
}
```

This approach supports a more principal-oriented `AccessSpec` definition approach:
```typescript
let accessSpecs: Array<AccessSpec> = [
  {
      accessCapabilities: new Set([AccessCapability.AdministerResource, AccessCapability.ReadConfig]),
      allowPrincipalArns: new Set(['admin1', 'cicd']),
      test: "ArnEquals"
  },

  {
      accessCapabilities: new Set([AccessCapability.ReadConfig]),
      allowPrincipalArns: new Set(['auditor', 'observability']),
      test: "ArnEquals"
  },
  {
      accessCapabilities: new Set([AccessCapability.ReadData, AccessCapability.WriteData]),
      allowPrincipalArns: new Set(['app1']),
      test: "ArnLike"
  }
];
```

Added a number of tests to cover the merge process.

What do you think?

brentryan commented:
This looks good @skuenzli. This should cleanup the api a bit for consumers now that all the merging is taken care of internally by k9-cdk. Once a release is cut then I'll give it a try in our projects and report back any issues.

skuenzli commented:
Thanks @brentryan! Will update docs and merge today.
